plugins/outputs/application_insights/application_insights.go

```diff
@@ -0,0 +1,410 @@
+		a.InstrumentationKey = iKey
+	}
+
+	if a.InstrumentationKey != "" {
+		a.Transmitter = NewAppinsightsTransmitter(a.InstrumentationKey)
+	}
+}
+
+func canBeTreatedAsAggregate(metric telegraf.Metric) bool {
+	// If we can identify the Sum and Count properties, then we are good
+	hasSum := metric.HasField("value") || metric.HasField("sum")
+	hasCount := metric.HasField("count") || metric.HasField("samples")
+	return hasSum && hasCount
+}
+
+func createSimpleMetricTelemetry(metric telegraf.Metric) *appinsights.MetricTelemetry {
+	telemetry := appinsights.NewMetricTelemetry(metric.Name(), 0.0)
+	telemetry.Properties = metric.Tags()
+	telemetry.Timestamp = metric.Time()
+
+	err := setFloat64TelemetryProperty(&(telemetry.Value), []string{"value"}, metric, nil)
+	if err == nil {
+		addRemainingMeasurements(telemetry, metric, []string{"value"})
+	} else {
+		// Either the metric does not have a "value" field, or it was not convertable to float64
+		// Heuristics: if the metric has just one field that is convertible to float64, take it as value
+		// Otherwise give up
+		convertedFieldCount := 0
+		var metricValue float64
+
+		for _, value := range metric.Fields() {
+			convertedValue, err := toFloat64(value)
+			if err == nil {
+				convertedFieldCount++
+				metricValue = convertedValue
+			}
+		}
+
+		if convertedFieldCount != 1 {
+			return nil
+		} else {
+			telemetry.Value = metricValue
+		}
+	}
+
+	return telemetry
+}
+
+func createAggregateMetricTelemetry(metric telegraf.Metric) *appinsights.AggregateMetricTelemetry {
+	telemetry := appinsights.NewAggregateMetricTelemetry(metric.Name())
+	telemetry.Properties = metric.Tags()
+	telemetry.Timestamp = metric.Time()
+
+	var usedFields []string
+	var err error
+
+	// Get the sum of all individual measurements(mandatory property)
+	err = setFloat64TelemetryProperty(&(telemetry.Value), []string{"sum", "value"}, metric, &usedFields)
+	if err != nil {
+		return nil
+	}
+
+	// Get the count of measurements (mandatory property)
+	err = setIntTelemetryProperty(&(telemetry.Count), []string{"count", "samples"}, metric, &usedFields)
+	if err != nil {
+		return nil
+	}
+
+	// We attempt to set Min, Max, StdDev and Variance fields but do not really care if they are not present
+	setFloat64TelemetryProperty(&(telemetry.Min), []string{"min"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Max), []string{"max"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.StdDev), []string{"stddev"}, metric, &usedFields)
+	setFloat64TelemetryProperty(&(telemetry.Variance), []string{"variance"}, metric, &usedFields)
+
+	addRemainingMeasurements(telemetry, metric, usedFields)
+
+	return telemetry
+}
+
+func setFloat64TelemetryProperty(
+	telemetryProperty *float64,
+	candidateFields []string,
+	metric telegraf.Metric,
+	usedFields *[]string) error {
+
+	for _, fieldName := range candidateFields {
+		if metric.HasField(fieldName) {
+			metricValue, err := toFloat64(metric.Fields()[fieldName])
```

Comment:
What if the fieldName afterwards can be converted? Do we still want to return err?

Reply:
@yantang-msft Not sure I understand the question. The error is about inability to convert to float64 and has nothing to do with the field name.

Reply:
I mean we have a list of candidateFields. If the first candidate field failed to convert, but the second candidate field can be converted, do we want to return as soon as there is an error or wait after all fields are checked.

Reply:
Ah, good point--we should continue. Thanks